File: bindings/dart/example/hyperliquid_provider.dart
import 'package:maxt/maxt.dart';

const _address = String.fromEnvironment('HYPERLIQUID_ADDRESS');

Future<void> main() async {
  await Maxt.initialize();
  try {
    final market = Market.perpetual(Exchange.hyperliquid, 'BTC', 'USDC');
    final adapter = HyperliquidAdapter();
    final mids = await adapter.allMids();
    final book = await adapter.l2Book(market);
    final trades = await adapter.recentTrades(market);
    print(
      '${mids.length} mid prices; ${book.bids.length + book.asks.length} levels; ${trades.length} trades',
    );

    if (_address.isEmpty) {
      print('Pass -DHYPERLIQUID_ADDRESS=0x... for address-scoped Info reads.');
    } else {
      final orders = await HyperliquidAdapter(
        address: _address,
      ).basicOpenOrders();
      print('${orders.length} address-scoped open orders');
    }
  } finally {
    await Maxt.dispose();
  }
}

File: examples/hyperliquid_provider.rs
//! Read Hyperliquid public market data and optionally address-scoped Info data.
//!
//! ```text
//! cargo run --example hyperliquid_provider
//! HYPERLIQUID_ADDRESS=0x... cargo run --example hyperliquid_provider
//! ```

use maxt::adapters::HyperliquidAdapter;
use maxt::{Exchange, Market};

#[tokio::main(flavor = "current_thread")]
async fn main() -> maxt::Result<()> {
    let market = Market::perpetual(Exchange::Hyperliquid, "BTC", "USDC");
    let adapter = HyperliquidAdapter::new();
    let mids = adapter.all_mids().await?;
    let book = adapter.l2_book(&market).await?;
    let trades = adapter.recent_trades(&market).await?;
    println!(
        "{} mid prices; {} book levels; {} recent trades",
        mids.len(),
        book.bids.len() + book.asks.len(),
        trades.len()
    );

    if let Ok(address) = std::env::var("HYPERLIQUID_ADDRESS") {
        let account = HyperliquidAdapter::new().with_query_address(address);
        let orders = account.basic_open_orders().await?;
        println!("{} address-scoped open orders", orders.len());
    } else {
        println!(
            "set HYPERLIQUID_ADDRESS to read an address-scoped Info response without a private key"
        );
    }
    Ok(())
}
